mlrun/api/crud/secrets.py

```diff
@@ -182,5 +203,185 @@ def list_secrets(
+                        f"Storing secret using key map is not implemented for provider {secrets.provider}"
+                    )
+                if not self._is_key_map_secret_key(key_map_secret_key):
+                    raise mlrun.errors.MLRunInvalidArgumentError(
+                        f"Key map secret key must start with: {self.key_map_secrets_key_prefix}"
+                    )
+                if not allow_internal_secrets:
+                    raise mlrun.errors.MLRunAccessDeniedError(
+                        f"Not allowed to create/update internal secrets (key starts with "
+                        f"{self.internal_secrets_key_prefix})"
+                    )
+                self.validate_secret_key_regex(key_map_secret_key)
+                secrets_to_store_in_key_map = [
+                    secret_key
+                    for secret_key in secrets_to_store.keys()
+                    if self._is_secret_stored_in_key_map(secret_key)
+                ]
+                if secrets_to_store_in_key_map:
+                    key_map = (
+                        self._get_secret_key_map(project, key_map_secret_key) or {}
+                    )
+                    key_map.update(
+                        {
+                            secret_key: self._generate_uuid()
+                            for secret_key in secrets_to_store_in_key_map
+                            if secret_key not in key_map
+                        }
+                    )
+                    updated_secrets_to_store = {}
+                    for key, value in secrets_to_store.items():
+                        if key in secrets_to_store_in_key_map:
+                            updated_secrets_to_store[key_map[key]] = value
+                        else:
+                            updated_secrets_to_store[key] = value
+                    updated_secrets_to_store[key_map_secret_key] = json.dumps(key_map)
+                    secrets_to_store = updated_secrets_to_store
+        return secrets_to_store
+
+    def _get_secret_key_map(
+        self, project: str, key_map_secret_key: str,
+    ) -> typing.Optional[dict]:
+        secrets_data = self.list_secrets(
+            project,
+            mlrun.api.schemas.SecretProviderName.kubernetes,
+            [key_map_secret_key],
+            allow_secrets_from_k8s=True,
+            allow_internal_secrets=True,
+        )
+        value = secrets_data.secrets.get(key_map_secret_key)
+        if value:
+            value = json.loads(value)
+        return value
+
+    def _is_secret_stored_in_key_map(self, key: str) -> bool:
+        # Key map are only used for invalid keys
+        return not self.validate_secret_key_regex(key, raise_on_failure=False)
+
+    def _is_internal_secret_key(self, key: str) -> bool:
         return key.startswith(self.internal_secrets_key_prefix)
+
+    def _is_key_map_secret_key(self, key: str) -> bool:
+        return key.startswith(self.key_map_secrets_key_prefix)
```

Comment:
```suggestion
        return key.startswith(self.key_map_secrets_key_prefix) and validate_secret_key_regex(key, raise_on_failure=False)
```
A key-map secret must be stored as a 1st level secret (otherwise you'll get issues with key-map pointing at key-map which obviously won't work).

Reply:
I already the validate the key map secret key regex on creation (in `_validate_and_enrich_secrets_to_store` in line 349) as a proof, the test you asked to verify that was working
I don't see a reason validating the key every time I'm calling this function which is just there to check whether we're dealing with a key map secret key

Reply:
If the client is used a key name generated by `generate_schedule_key_map_secret_key` then of course there's no issue. But, the other methods actually get a key-map name, which can be anything. For example, I can call `get_secret` and pass a name that is not valid for the key-map. In this case the flow will eventually fail in `_resolve_secret_key` as it tries to `list_secrets` on the invalid key name. It's not terrible, but it is easier to debug if there's a specific check for the key-map name that puts out an indicative error on that.
I don't mind too much keeping it as-is, but it would have been clearer that way.